File: Week 2/Homework/hangman.swift
import Foundation
let words = ["PIZZA", "CHEESE", "ICE", "CHICKEN", "JAZZ", "ABRUPTLY", "COMPUTER", "BOARD", "PICTURE", "GALAXY", "FIXABLE", "JIGSAW", "INJURY", "KAZOO", "KEYHOLE", "WIZARD", "YOUTHFUL", "WALKWAY", "DOG", "CAT"]
let randomWord:String = words[Int(arc4random_uniform(20) + 1)]
var letters:[String] = []
var guessedString = ""
for letter in Array(randomWord.characters){
	guessedString = guessedString + " _";
	letters.append(String(letter))
}
letters = Array(Set(letters))

func getIndexesOfLetter(word:String, letter:String) -> [Int]{
	var indexes:[Int] = []
	var i = 0
	for character in word.characters{
		if String(character) == letter{
			indexes.append(i)
		}
		i += 1
	}
	return indexes
}

func arrayToString(array:[String]) -> String{
	var finalString = ""
	var index = 0
	for string in array{
		if (index == 0) && (array.count == 1 || array.count == 2){
			finalString = finalString + string + " "
		}
		else if (index == array.count - 1){
			finalString = finalString + "and " + string
		}
		else{
			finalString = finalString + string + ", "
		}
		index += 1
	}
	return finalString
}

func game(word:String, guessedString:String, letters:[String], guesses:Int, guessedLetters:[String]){
	var newGuessedString = guessedString
	var newGuesses = guesses
	var newGuessedLetters = guessedLetters
	print(guessedString)
	print(" ")
	if(guessedLetters.count > 0){
		print("Guessed Letters: " + arrayToString(array:guessedLetters))
	}
	print("Guess a letter: ", terminator: "")
	var guess = String(readLine()!)!
	let index = guess.index(guess.startIndex, offsetBy: 0)
	guess = String(guess[index]).uppercased()
	if letters.contains(guess){
		print(" ")
		print("Correct!")
		let indexes = getIndexesOfLetter(word:word, letter:guess)
		var charArray = Array(guessedString.characters)
		for i in indexes {
			let ii = Int((i)*2 + 1)
			charArray[ii] = Character(guess)
		}
		newGuessedString = String(charArray)
		newGuessedLetters.append(guess)
		if Array(newGuessedString.characters).contains(Character("_")){
			game(word:word, guessedString:newGuessedString, letters:letters, guesses:newGuesses, guessedLetters:newGuessedLetters)
		}
		else{
			print("Congratulations! You Won!")
			print(" ")
			print(newGuessedString)
		}
	}
	else{
		newGuessedLetters.append(guess)
		print(" ")
		newGuesses -= 1
		print("Wrong Guess!")
		if newGuesses == 0{
			print("Game Over! The word was " + word)
		}
		else{
			if newGuesses > 1{
				print("You have " + String(newGuesses) + " more guesses!")
			}
			else{
				print("You have 1 more guess!")
			}
			game(word:word, guessedString:newGuessedString, letters:letters, guesses:newGuesses, guessedLetters:newGuessedLetters)
		}
	}
}

game(word:randomWord, guessedString:guessedString, letters:letters, guesses:5, guessedLetters:[])

File: Week 2/Homework/Hangman.java
import java.util.*;
import java.util.Scanner;
import java.util.concurrent.ThreadLocalRandom;
public class Hangman{
	public static void main(String[] args) {
		String[] words = new String[20];
		words[0] = "PIZZA";
		words[1] = "CHEESE";
		words[2] = "ICE";
		words[3] = "CHICKEN";
		words[4] = "JAZZ";
		words[5] = "ABRUPTLY";
		words[6] = "COMPUTER";
		words[7] = "BOARD";
		words[8] = "PICTURE";
		words[9] = "GALAXY";
		words[10] = "FIXABLE";
		words[11] = "JIGSAW";
		words[12] = "INJURY";
		words[13] = "KAZOO";
		words[14] = "KEYHOLE";
		words[15] = "WIZARD";
		words[16] = "YOUTHFUL";
		words[17] = "WALKWAY";
		words[18] = "DOG";
		words[19] = "CAT";
		String randomWord = words[(ThreadLocalRandom.current().nextInt(0, 20))];
		char[] characters = randomWord.toCharArray();
		String guessedString = "";
		ArrayList avalableLetters = new ArrayList();
		for (char character : characters) {
			guessedString = guessedString + " _";
			if(!avalableLetters.contains(String.valueOf(character))){
				avalableLetters.add(String.valueOf(character));
			}
		}
		game(randomWord, guessedString, avalableLetters, 10, new ArrayList());
	}
	public static void game(String word, String guessedString, ArrayList letters, int guesses, ArrayList guessedLetters){
		Scanner scanner = new Scanner(System.in);
		System.out.println(" ");
		System.out.println(guessedString);
		System.out.println(" ");
		if(guessedLetters.size() >= 1){
			String guessedLettersString = listToString(guessedLetters);
			System.out.println("Current Guesses: " + guessedLettersString);
		}
		System.out.print("Guess a letter: ");
		String letter = String.valueOf(scanner.next().replaceAll("[^a-zA-Z]", "").toCharArray()[0]).toUpperCase();
		guessedLetters.add(letter);
		if(letters.contains(letter)){
			ArrayList indexes = getIndexesOfLetter(word, letter);
			for(int i = 0; i < indexes.size(); i++) {
				int index = Integer.valueOf(String.valueOf(indexes.get(i))) * 2 + 1;
				guessedString = guessedString.substring(0, index) + letter + guessedString.substring(index + 1);
			}
			System.out.println(" ");
			System.out.println("Correct!");
			if(didWin(guessedString)){
				System.out.println("Congatulations! You Won!");
				System.out.println(guessedString);
				System.out.println(" ");
			}
			else{
				game(word, guessedString, letters, guesses, guessedLetters);
			}
		}
		else{
			String guessedLettersString = listToString(guessedLetters);
			System.out.println(" ");
			System.out.println("Wrong Guess!");
			if (guesses == 1){
				System.out.println("Game Over! The word was " + word);
			}
			else{
				if (guesses == 2) {
					System.out.println("You have 1 more guess!");
				}
				else{
					System.out.println("You have " + (guesses - 1)  +  " more guesses!");
				}
				game(word, guessedString, letters, (guesses - 1), guessedLetters);
			}
		}
	}

	public static ArrayList getIndexesOfLetter(String word, String letter){
		char[] characters = word.toCharArray();
		ArrayList indexes = new ArrayList();
		int i = 0;
		for (char character : characters) {
			if(character == letter.toCharArray()[0]){
				indexes.add(i);
			}
			i = i + 1;
		}
		return indexes;
	}

	public static boolean didWin(String guessedString){
		boolean didWin = true;
		char[] characters = guessedString.toCharArray();
		for (char character : characters) {
			if (character == "_".toCharArray()[0]){
				didWin = false;
			}
		}
		return didWin;
	}

	public static String listToString(ArrayList list){
		String finalString = "";
		for(int i = 0; i < list.size(); i++){
			if(i == list.size() - 1 && list.size() > 1){
				finalString = finalString + String.valueOf(list.get(i));
			}
			else if(list.size() == 1 || list.size() == 2){
				finalString = finalString + String.valueOf(list.get(i)) + " ";
			}
			else{
				finalString = finalString + String.valueOf(list.get(i)) + ", ";
			}
		}
		return finalString;
	}
}
